src/latency.c

```diff
@@ -490,6 +490,147 @@ sds createLatencyReport(void) {
+    struct hdr_iter iter;
+    hdr_iter_log_init(&iter, histogram, 1024,2);
+    int64_t previous_count = 0;
+    while (hdr_iter_next(&iter)) {
+        const int64_t micros = iter.highest_equivalent_value / 1000;
+        const int64_t cumulative_count = iter.cumulative_count;
+        if(cumulative_count > previous_count){
+            addReplyLongLong(c,(long long) micros);
+            addReplyLongLong(c,(long long) cumulative_count);
+            samples+=2;
+        }
+        previous_count = cumulative_count;
+    }
+    setDeferredArrayLen(c,replylen,samples);
+}
+
+/* latencyCommand() helper to produce a per command category cumulative distribution of latencies. */
+void latencyAllCommandsFillCDF(client *c) {
+    dictIterator *di = dictGetSafeIterator(server.commands);
+    dictEntry *de;
+    struct redisCommand *cmd;
+    sds info = sdsempty();
+    while((de = dictNext(di)) != NULL) {
+        char *tmpsafe;
+        cmd = (struct redisCommand *) dictGetVal(de);
+        if (!cmd->latency_histogram)
+            continue;
+        info = fillCommandCDF(info, getSafeInfoString(cmd->name, strlen(cmd->name), &tmpsafe), cmd->latency_histogram);
+        if (tmpsafe != NULL) zfree(tmpsafe);
+    }
+    dictReleaseIterator(di);
+    addReplyVerbatim(c,info,sdslen(info),"txt");
+    sdsfree(info);
+}
+
+/* latencyCommand() helper to produce a per command category cumulative distribution of latencies. */
+void latencyCommandFillCDF(client *c) {
+    sds info = sdsempty();
+    for (int j = 2; j < c->argc; j++){
+        struct redisCommand *cmd = dictFetchValue(server.commands, c->argv[j]->ptr);
+        if (cmd == NULL) {
+            addReplyErrorFormat(c,
+                "Invalid command name '%s'", (char*) c->argv[j]->ptr);
+            sdsfree(info);
+            return;
+        }
+        char *tmpsafe;
+        if (!cmd->latency_histogram)
+            continue;
+        info = fillCommandCDF(info, getSafeInfoString(cmd->name, strlen(cmd->name), &tmpsafe), cmd->latency_histogram);
+        if (tmpsafe != NULL) zfree(tmpsafe);
+    }
+    addReplyVerbatim(c,info,sdslen(info),"txt");
+    sdsfree(info);
+}
+
+/* latencyCommand() helper to produce a per command category cumulative distribution of latencies in RESP format. */
+void latencyAllCommandsFillCDFResp(client *c) {
+    dictIterator *di = dictGetSafeIterator(server.commands);
+    dictEntry *de;
+    struct redisCommand *cmd;
+    void *replylen = addReplyDeferredLen(c);
+    int command_with_data = 0;
+    while((de = dictNext(di)) != NULL) {
+        char *tmpsafe;
+        cmd = (struct redisCommand *) dictGetVal(de);
+        if (!cmd->latency_histogram)
+            continue;
+        fillCommandCDFResp(c, getSafeInfoString(cmd->name, strlen(cmd->name), &tmpsafe), cmd->latency_histogram);
+        if (tmpsafe != NULL) zfree(tmpsafe);
+        command_with_data++;
+    }
+    dictReleaseIterator(di);
+    setDeferredArrayLen(c,replylen,command_with_data);
```

Comment:
shouldn't that one be a map (command name is the field, and the array of histogram entries the value)?
in that case, we should do the same on the other variant (the one that takes command names as input).

p.s. i think we can make these two share more code in some way.
i.e. they're both doing exactly the same, with the exception of the loop.
i don't have a concrete suggestion. just the desire to find something.

Reply:
> shouldn't that one be a map (command name is the field, and the array of histogram entries the value)?

yes indeed. Made the change on the latest commit. New look:

```
$ ./src/redis-cli -3 latency histogram set get
1# "set" => 1# "calls" => (integer) 1000
   2# "histogram_usec" => 1# (integer) 1 => (integer) 996
      2# (integer) 2 => (integer) 998
      3# (integer) 4 => (integer) 999
      4# (integer) 8 => (integer) 1000
2# "get" => 1# "calls" => (integer) 1000
   2# "histogram_usec" => 1# (integer) 1 => (integer) 990
      2# (integer) 2 => (integer) 993
      3# (integer) 4 => (integer) 996
      4# (integer) 8 => (integer) 998
      5# (integer) 16 => (integer) 999
      6# (integer) 33 => (integer) 1000
$ ./src/redis-cli -3 latency histogram-report set get
latency_hist_usec_set:calls=1000,histogram=[1=996,2=998,4=999,8=1000]
latency_hist_usec_get:calls=1000,histogram=[1=990,2=993,4=996,8=998,16=999,33=1000]
```

Reply:
Actually, redis-cli shows the resp one pretty readable. 
Maybe we can drop the human subcommand? 
The only disadvantage is that there are multiple lines ER command, but it is actually more readable that the one liner... 
Maybe there's an advantage to have one that outputs CSV? I guess that's the responsibility of redis-cli to convert (we have the new `--json`). 

I'm leaning towards dropping the human format.

Reply:
> I'm leaning towards dropping the human format.

the less code the better IMHO. Lets drop it :)